src/iidm/ShuntCompensatorNonLinearModel.cpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) 2020, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <powsybl/iidm/ShuntCompensatorNonLinearModel.hpp>
+
+#include <boost/range/adaptor/transformed.hpp>
+
+#include <powsybl/iidm/ShuntCompensator.hpp>
+#include <powsybl/iidm/ValidationException.hpp>
+#include <powsybl/iidm/ValidationUtils.hpp>
+
+namespace powsybl {
+
+namespace iidm {
+
+ShuntCompensatorNonLinearModel::Section::Section(double b, double g) :
+    m_b(b),
+    m_g(g) {
+}
+
+double ShuntCompensatorNonLinearModel::Section::getB() const {
+    return m_b;
+}
+
+double ShuntCompensatorNonLinearModel::Section::getG() const {
+    return m_g;
+}
+
+ShuntCompensatorNonLinearModel::Section& ShuntCompensatorNonLinearModel::Section::setB(double b) {
+    checkB(m_model.get().getShuntCompensator(), b);
+    m_b = b;
+    return *this;
+}
+
+ShuntCompensatorNonLinearModel::Section& ShuntCompensatorNonLinearModel::Section::setG(double g) {
+    checkG(m_model.get().getShuntCompensator(), g);
+    m_g = g;
+    return *this;
+}
+
+void ShuntCompensatorNonLinearModel::Section::setModel(ShuntCompensatorNonLinearModel& model) {
+    m_model = stdcxx::ref(model);
+}
+
+ShuntCompensatorNonLinearModel::ShuntCompensatorNonLinearModel(ShuntCompensator& shuntCompensator, const std::vector<Section>& sections) :
+    ShuntCompensatorModel(shuntCompensator),
+    m_sections(sections) {
+    for (Section& section : m_sections) {
+        section.setModel(*this);
+    }
+}
+
+stdcxx::const_range<ShuntCompensatorNonLinearModel::Section> ShuntCompensatorNonLinearModel::getAllSections() const {
+    const auto& mapper = [](const Section& section) -> const Section& {
+        return section;
+    };
+    return m_sections | boost::adaptors::transformed(mapper);
+}
+
+stdcxx::range<ShuntCompensatorNonLinearModel::Section> ShuntCompensatorNonLinearModel::getAllSections() {
+    const auto& mapper = [](Section& section) -> Section& {
+        return section;
+    };
+    return m_sections | boost::adaptors::transformed(mapper);
+}
```

Comment:
I don't understand why do you need to transform the sections. You should be able to just return `m_sections`